File: sdk/python/src/inferlet/media.py
"""
Multimodal media wrappers over ``pie:core/media``.
"""

from __future__ import annotations

from typing import Iterable

from wit_world.imports import media as _media

from .model import Model


def _as_bytes(data: bytes | bytearray | memoryview | Iterable[int]) -> bytes:
    if isinstance(data, bytes):
        return data
    return bytes(data)


class Image:
    """Host-side preprocessed still image."""

    __slots__ = ("_handle",)

    def __init__(self, handle: _media.Image) -> None:
        self._handle = handle

    def __enter__(self) -> Image:
        """Return this wrapper while keeping the underlying resource alive."""
        return self

    def __exit__(self, exc_type, exc_value, traceback):
        """Release the underlying image resource."""
        return self._handle.__exit__(exc_type, exc_value, traceback)

    @classmethod
    def from_bytes(
        cls,
        model: Model,
        data: bytes | bytearray | memoryview | Iterable[int],
    ) -> Image:
        """Decode image bytes for ``model``."""
        return cls(_media.Image.from_bytes(model._handle, _as_bytes(data)))

    def token_count(self) -> int:
        """Soft-token rows / KV slots occupied by this span."""
        return self._handle.token_count()

    def position_span(self) -> int:
        """Sequence-position span consumed by this image."""
        return self._handle.position_span()

    def grid(self) -> tuple[int, int, int]:
        """``(t, h, w)`` merged-token grid."""
        return self._handle.grid()

    def prefix_tokens(self) -> list[int]:
        """Model-specific delimiter tokens before the image span."""
        return list(self._handle.prefix_tokens())

    def suffix_tokens(self) -> list[int]:
        """Model-specific delimiter tokens after the image span."""
        return list(self._handle.suffix_tokens())


class Video:
    """Host-side decoded and sampled video clip."""

    __slots__ = ("_handle",)

    def __init__(self, handle: _media.Video) -> None:
        self._handle = handle

    def __enter__(self) -> Video:
        """Return this wrapper while keeping the underlying resource alive."""
        return self

    def __exit__(self, exc_type, exc_value, traceback):
        """Release the underlying video resource."""
        return self._handle.__exit__(exc_type, exc_value, traceback)

    @classmethod
    def from_bytes(
        cls,
        model: Model,
        data: bytes | bytearray | memoryview | Iterable[int],
        max_frames: int,
    ) -> Video:
        """Decode and uniformly sample up to ``max_frames`` frames."""
        return cls(_media.Video.from_bytes(model._handle, _as_bytes(data), max_frames))

    def frame_count(self) -> int:
        """Number of sampled frames."""
        return self._handle.frame_count()

    def frame(self, index: int) -> Image:
        """The ``index``-th sampled frame as an image span."""
        return Image(self._handle.frame(index))

    def timestamp(self, index: int) -> float:
        """Timestamp in seconds for the ``index``-th sampled frame."""
        return self._handle.timestamp(index)


class Audio:
    """Host-side preprocessed audio clip."""

    __slots__ = ("_handle",)

    def __init__(self, handle: _media.Audio) -> None:
        self._handle = handle

    def __enter__(self) -> Audio:
        """Return this wrapper while keeping the underlying resource alive."""
        return self

    def __exit__(self, exc_type, exc_value, traceback):
        """Release the underlying audio resource."""
        return self._handle.__exit__(exc_type, exc_value, traceback)

    @classmethod
    def from_bytes(
        cls,
        model: Model,
        data: bytes | bytearray | memoryview | Iterable[int],
    ) -> Audio:
        """Decode audio bytes for ``model``."""
        return cls(_media.Audio.from_bytes(model._handle, _as_bytes(data)))

    def token_count(self) -> int:
        """Soft-token rows / KV slots occupied by this clip."""
        return self._handle.token_count()

    def position_span(self) -> int:
        """Sequence-position span consumed by this audio clip."""
        return self._handle.position_span()

    def prefix_tokens(self) -> list[int]:
        """Model-specific delimiter tokens before the audio span."""
        return list(self._handle.prefix_tokens())

    def suffix_tokens(self) -> list[int]:
        """Model-specific delimiter tokens after the audio span."""
        return list(self._handle.suffix_tokens())

File: sdk/inferlet/javascript/src/media.ts
// Multimodal input — `pie:inferlet/media`.
//
// The inferlet hands the host raw encoded bytes (`Image.fromBytes` for
// PNG/JPEG, `Video.fromBytes` for animated GIF, `Audio.fromBytes` for WAV)
// and the host decodes + preprocesses per the bound model. A span enters the
// sequence as the token run its handle answers, and the handle crosses again
// beside the tokens (`fwd.media([{ tag: 'image', val: img.handle }])`).

import * as _media from 'pie:inferlet/media@0.3.0';

export class MediaError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'MediaError';
  }
}

function wit<T>(what: string, f: () => T): T {
  try {
    return f();
  } catch (e: unknown) {
    const payload = (e as { payload?: unknown })?.payload;
    throw new MediaError(`${what}: ${typeof payload === 'string' ? payload : String(e)}`);
  }
}

export interface MergedGrid {
  t: number;
  h: number;
  w: number;
}

export class Image {
  /** @internal */
  constructor(readonly handle: _media.Image) {}

  static fromBytes(data: Uint8Array): Image {
    return new Image(wit('image', () => _media.Image.fromBytes(data)));
  }

  /** The placeholder run this span enters the sequence as. */
  tokens(): Uint32Array {
    return this.handle.tokens();
  }
  digest(): Uint8Array {
    return this.handle.digest();
  }
  tokenCount(): number {
    return this.handle.tokenCount();
  }
  positionSpan(): number {
    return this.handle.positionSpan();
  }
  grid(): MergedGrid {
    return this.handle.grid();
  }
  prefixTokens(): Uint32Array {
    return this.handle.prefixTokens();
  }
  suffixTokens(): Uint32Array {
    return this.handle.suffixTokens();
  }
}

export class Audio {
  /** @internal */
  constructor(readonly handle: _media.Audio) {}

  static fromBytes(data: Uint8Array): Audio {
    return new Audio(wit('audio', () => _media.Audio.fromBytes(data)));
  }

  tokens(): Uint32Array {
    return this.handle.tokens();
  }
  digest(): Uint8Array {
    return this.handle.digest();
  }
  tokenCount(): number {
    return this.handle.tokenCount();
  }
  positionSpan(): number {
    return this.handle.positionSpan();
  }
  prefixTokens(): Uint32Array {
    return this.handle.prefixTokens();
  }
  suffixTokens(): Uint32Array {
    return this.handle.suffixTokens();
  }
}

export class Video {
  /** @internal */
  constructor(readonly handle: _media.Video) {}

  static fromBytes(data: Uint8Array, maxFrames: number): Video {
    return new Video(wit('video', () => _media.Video.fromBytes(data, maxFrames)));
  }

  frameCount(): number {
    return this.handle.frameCount();
  }
  frame(index: number): Image {
    return new Image(wit(`video frame ${index}`, () => this.handle.frame(index)));
  }
  timestamp(index: number): number {
    return this.handle.timestamp(index);
  }
}
